packages/compiler/src/core/checker.ts

```diff
@@ -2988,6 +3665,496 @@ export function createChecker(program: Program): Checker {
+            format: { actual: node.arguments.length.toString(), expected },
+            target: node,
+          })
+        );
+      }
+    }
+
+    const resolvedArgs: Value[] = [];
+
+    for (const [index, parameter] of declaration.parameters.entries()) {
+      if (parameter.rest) {
+        const restType = getIndexType(parameter.type);
+        if (restType) {
+          for (let i = index; i < node.arguments.length; i++) {
+            const argNode = node.arguments[i];
+            if (argNode) {
+              const arg = getValueForNode(argNode, mapper, { kind: "argument", type: restType });
+              if (arg === null) {
+                hasError = true;
+                continue;
+              }
+              if (checkValueOfType(arg, restType, argNode)) {
+                resolvedArgs.push(arg);
+              } else {
+                hasError = true;
+              }
+            }
+          }
+        }
+        break;
+      }
+      const argNode = node.arguments[index];
+      if (argNode) {
+        const arg = getValueForNode(argNode, mapper, {
+          kind: "argument",
+          type: parameter.type as any, // TODO: change if we change this to not be a FunctionParameter
+        });
+        if (arg === null) {
+          hasError = true;
+          continue;
+        }
+        if (checkValueOfType(arg, parameter.type as any, argNode)) {
+          resolvedArgs.push(arg);
+        } else {
+          hasError = true;
+        }
+      }
+    }
+    if (hasError) {
+      return null;
+    }
+    return {
+      valueKind: "ScalarValue",
+      value: {
+        name: declaration.name,
+        args: resolvedArgs,
+      },
+      scalar: declaration.scalar,
+      type: declaration.scalar,
+    };
+  }
+
+  // TODO: should those be called eval?
+  function checkCallExpression(
+    node: CallExpressionNode,
+    mapper: TypeMapper | undefined
+  ): Value | null {
+    const target = checkCallExpressionTarget(node, mapper);
+    if (target === null) {
+      return null;
+    }
+    if (target.kind === "ScalarConstructor") {
+      return createScalarValue(node, mapper, target);
+    }
+
+    if (areScalarsRelated(target, getStdType("string"))) {
+      return checkPrimitiveArg(node, target, "StringValue");
+    } else if (areScalarsRelated(target, getStdType("numeric"))) {
+      return checkPrimitiveArg(node, target, "NumericValue");
+    } else if (areScalarsRelated(target, getStdType("boolean"))) {
+      return checkPrimitiveArg(node, target, "BooleanValue");
+    } else {
+      reportCheckerDiagnostic(
+        createDiagnostic({
+          code: "named-init-required",
+          format: { typeKind: target.kind },
+          target: node.target,
+        })
+      );
+      return null;
+    }
+  }
+
+  function checkTypeOfExpression(node: TypeOfExpressionNode, mapper: TypeMapper | undefined): Type {
+    const entity = checkNode(node.target, mapper, undefined);
```

Comment:
can just call getTypeForNode?

Reply:
this doesn't work we wnat to get the type of a value not just figure out if indeterminate should be a type